src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,129 @@
+
+    public static propTypes = {
+        children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
+        className: PropTypes.string,
+        id: PropTypes.string,
+        isOpen: PropTypes.bool,
+        onDismiss: PropTypes.func,
+        title: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node, PropTypes.string]).isRequired
+    };
+
+    private eventHandler;
+
+    constructor(props) {
+        super(props);
+        this.eventHandler = ModalEvent.handler(this.handleClickOutside);
+        this.handleEventRegistration = this.handleEventRegistration.bind(this);
+    }
+
+    public componentDidMount() {
+        if (this.props.onDismiss) {
+            this.handleEventRegistration();
+        }
+    }
+
+    public componentDidUpdate(prevProps) {
+        if(this.props.onDismiss !== prevProps.onDismiss || this.props.isOpen !== prevProps.isOpen) {
+            this.handleEventRegistration();
+        }
+    }
+
+    public componentWillUnmount() {
+        this.eventHandler.deregister();
+    }    
+    
+    public render() {
+        const { children, className, id = Rivet.shortuid(), isOpen, onDismiss, title, ...attrs } = this.props;
+        return (
+            <div
+                className={classNames(["rvt-modal", className])}
+                id={id}
+                role="dialog"
+                aria-labelledby={`${id}-title`}
+                aria-hidden={!isOpen}
+                tabIndex={-1}
+                {...attrs}
+            >
+                <div className="rvt-modal__inner">
+                    <header className="rvt-modal__header">
+                        <h1 className="rvt-modal__title" id={`${id}-title`}>{title}</h1>
+                        { onDismiss && <ModalCloseButton onDismiss={onDismiss} /> }
+                    </header>
+                    { children }
+                </div>
+            </div>
+        ); 
+    }
+
+
+    private closeModal = (event) => {
+        this.props.onDismiss && this.props.onDismiss();        
+    }
+
+    private shouldToggleModal = (event: ModalEvent) => {
+        if (event.isRightMouseClick() || event.isUnhandledKeyPress()) {
+            // If the user right clicks anywhere on the screen or they press an unhandled key do not close the menu
+            return false;
+        }
+
+        if (event.targets(ReactDOM.findDOMNode(this)) && !event.isKeyEvent()) {
+            // If the user clicks or touches inside the modal do not close the menu
+            return false;
+        }
+
+        return true;
+    }
+
+    private handleClickOutside = (event: ModalEvent) => {
+        if(event && this.shouldToggleModal(event)) {
+            this.closeModal(event);
+        }
+    }    
+
+    protected handleEventRegistration() {
+        if(this.props.onDismiss && this.props.isOpen) {
+            this.eventHandler.register();
+        } else {
+            this.eventHandler.deregister();
+        }
+    }    
+
+}
```

Comment:
I'm wondering if we could expose the `ModalBody` and `ModalControls` as `Modal.Body` and `Modal.Controls` like we're doing with the header?

Reply:
I'll give it a shot. I hadn't noticed we'd found a way to do that.